### PR TITLE
[SPARK-9803] [SPARKR] Add subset and transform + tests

### PR DESCRIPTION
**Pull request by felixcheung**
Add subset and transform
Also reorganize `[` & `[[` to subset instead of select

Note: for transform, transform is very similar to mutate. Spark doesn't seem to replace existing column with the name in mutate (ie. `mutate(df, age = df$age + 2)` - returned DataFrame has 2 columns with the same name 'age'), so therefore not doing that for now in transform.
Though it is clearly stated it should replace column with matching name (should I open a JIRA for mutate/transform?)

### CODE REVIEW

**Comment by felixcheung** [comment]
@shivaram

**Comment by shivaram** [comment]
Jenkins, ok to test

**Comment by shivaram** [comment]
cc @sun-rui , @yu-iskw (especially for the rdname changes)

**Comment by felixcheung** [comment]
Hmm, I'm guessing the generic is breaking signature matching somehow, with
`result <- transform(Ozone, logOzone = log(Ozone))`
I was running run-tests.sh locally and it passes. Though it looks like this test requires plyr - I'll re-run it then - can we make this a requirement in run-tests.sh?

**Comment by shivaram** [comment]
I don't think we should add more dependencies for tests and I don't see why we need plyr. The `transform` function in R is defined at https://stat.ethz.ch/R-manual/R-devel/library/base/html/transform.html and one thing is that the generic / method we define needs to exactly match the definition there (including the names of variables etc.)

**Comment by felixcheung** [comment]
This test doesn't run unless plyr is installed. Not sure why.
I think the mutate signature doesn't match either. I will update them.





On Fri, Aug 28, 2015 at 12:53 PM -0700, "Shivaram Venkataraman" <notifications@github.com> wrote:
I don't think we should add more dependencies for tests and I don't see why we need plyr. The `transform` function in R is defined at https://stat.ethz.ch/R-manual/R-devel/library/base/html/transform.html and one thing is that the generic / method we define needs to exactly match the definition there (including the names of variables etc.)

---

**Comment by shivaram** [comment]
mutate is only defined in `plyr` and not in base R, so we have less of a need to match its signature (though I guess it will be good to do that)

File: R/pkg/inst/tests/test_sparkSQL.R [hunk 1]
```diff
@@ -1042,6 +1046,19 @@ test_that("mutate(), rename() and names()", {
...
   expect_equal(names(newDF2)[1], "newerName")
+
+  transformedDF <- transform(df, newAge = -df$age, newAge2 = df$age / 2)
+  expect_equal(length(columns(transformedDF)), 4)
+  expect_equal(columns(transformedDF)[3], "newAge")
+  expect_equal(columns(transformedDF)[4], "newAge2")
+  expect_equal(first(filter(transformedDF, transformedDF$name == "Andy"))$newAge, -30)
+
+  # ensure the proper signature is used - otherwise this will fail to run
+  attach(airquality)
```
Comment (shivaram) [inline R/pkg/inst/tests/test_sparkSQL.R]: Hmm so this is testing whether `transform` on local data frames continue to work ? its an interesting idea -- could you update the comment on top to say `test if transform on local data frames works. ensures that proper signature ...`.

**Comment by shivaram** [comment]
@felixcheung for the column being replaced part, I'd say we can open another jira / discuss it for later. I prefer the existing behavior of appending columns and users can drop columns (either with a select or by setting it to `NULL` in SparkR)

cc @falaki

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -1034,6 +1068,10 @@ setMethod("[", signature(x = "DataFrame", i = "Column"),
 #' @return A new DataFrame with selected columns
 #' @export
 #' @rdname select
+#' @name select
+#' \seealso{
```
Comment (yu-iskw) [inline R/pkg/R/DataFrame.R]: roxygen2 has `@seealso` tag.

Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: should we use `@seealso` or `@family`?

File: R/pkg/R/generics.R [hunk 3]
```diff
@@ -467,7 +467,7 @@ setGeneric("merge")
 
 #' @rdname withColumn
 #' @export
-setGeneric("mutate", function(x, ...) {standardGeneric("mutate") })
+setGeneric("mutate", function(.data, ...) {standardGeneric("mutate") })
```
Comment (yu-iskw) [inline R/pkg/R/generics.R]: I'm not sure we should use `.` prefix for a parameter.

File: R/pkg/R/generics.R [hunk 4]
```diff
@@ -507,6 +507,10 @@ setGeneric("saveAsTable", function(df, tableName, source, mode, ...) {
   standardGeneric("saveAsTable")
 })
 
+#' @rdname withColumn
+#' @export
+setGeneric("transform", function(`_data`, ...) {standardGeneric("transform") })
```
Comment (yu-iskw) [inline R/pkg/R/generics.R]: I don't think we should use a quoted parameter. How about `data`?

File: R/pkg/R/generics.R (hunk 3 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: that's the signature from the original definition. please see discussion above.

File: R/pkg/R/generics.R (hunk 4 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: that's the signature from the original definition. please see discussion above.
local data.frame or vector will fail to run with transform if this doesn't match exactly.

Comment (yu-iskw) [inline R/pkg/R/generics.R]: Oh, I see.

File: R/pkg/R/generics.R (hunk 3 above)
Comment (yu-iskw) [inline R/pkg/R/generics.R]: I see. Thanks.

File: R/pkg/R/DataFrame.R (hunk 2 above)
Comment (yu-iskw) [inline R/pkg/R/DataFrame.R]: It would be better to use `@family`.

Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: added @family.

**Comment by felixcheung** [comment]
Opened an issue to track to replace behavior
https://issues.apache.org/jira/browse/SPARK-10346

**Comment by shivaram** [comment]
LGTM. @yu-iskw Any other comments ?

**Comment by yu-iskw** [comment]
LGTM

**Comment by shivaram** [comment]
Thanks @felixcheung @yu-iskw -- Merging this to master and branch-1.5

**Pull request closed without merging**